plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Stablecoin.hs

```diff
@@ -0,0 +1,385 @@
+                (bs
+                { bsStablecoins = bsStablecoins bs + sc'
+                , bsReserves = bsReserves bs + fmap round (fees + scValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scStablecoinTokenName (unSC sc'))
+            RedeemStablecoin sc' ->
+                let scValue = stablecoinNominalPrice bs rate * (BC $ fromInteger $ unSC sc') in
+                (bs
+                { bsStablecoins = bsStablecoins bs - sc'
+                , bsReserves = bsReserves bs + fmap round (fees - scValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scStablecoinTokenName (unSC $ negate sc'))
+            MintReserveCoin rc ->
+                let rcValue = reservecoinNominalPrice sc bs rate * (BC $ fromInteger $ unRC rc) in
+                (bs
+                { bsReservecoins = bsReservecoins bs + rc
+                , bsReserves = bsReserves bs + fmap round (fees + rcValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scReservecoinTokenName (unRC rc))
+            RedeemReserveCoin rc ->
+                let rcValue = reservecoinNominalPrice sc bs rate * (BC $ fromInteger $ unRC rc) in
+                (bs
+                { bsReservecoins = bsReservecoins bs - rc
+                , bsReserves = bsReserves bs + fmap round (fees - rcValue)
+                }, Constraints.mustForgeCurrency bsForgingPolicyScript scReservecoinTokenName (unRC $ negate rc))
+    guard $ isValidState sc newState rate
+    let dateConstraints = Constraints.mustValidateIn $ Interval.from obsSlot
+    pure (constraints <> newConstraints <> dateConstraints, newState)
+
+-- | A 'Value' with the given number of reservecoins
+reserveCoins :: Stablecoin -> Integer -> Value
+reserveCoins sc@Stablecoin{scReservecoinTokenName} =
+    let sym = Scripts.monetaryPolicyHash $ scriptInstance sc
+    in Value.singleton (Value.mpsSymbol sym) scReservecoinTokenName
+
+-- | A 'Value' with the given number of stablecoins
+stableCoins :: Stablecoin -> Integer -> Value
+stableCoins sc@Stablecoin{scStablecoinTokenName} =
+    let sym = Scripts.monetaryPolicyHash $ scriptInstance sc
+    in Value.singleton (Value.mpsSymbol sym) scStablecoinTokenName
+
+
+{-# INLINEABLE isValidState #-}
+isValidState :: Stablecoin -> BankState -> ConversionRate -> Bool
+isValidState sc bs@BankState{bsReservecoins, bsReserves, bsStablecoins} cr =
+    traceIfFalse "reservecoins < 0" (bsReservecoins >= RC 0)
+    && traceIfFalse "reserves < 0" (bsReserves >= BC 0)
+    && traceIfFalse "stablecoins < 0" (bsStablecoins >= SC 0)
+    && traceIfFalse "min reserves" (fmap fromInteger bsReserves >= minReserve sc cr bs)
+    && traceIfFalse "liabilities < 0" (liabilities bs cr >= zero)
+    && traceIfFalse "equity < 0" (equity bs cr >= zero)
+
+stablecoinStateMachine ::
+    Stablecoin
+    -> StateMachine BankState Input
+stablecoinStateMachine sc = StateMachine.mkStateMachine (transition sc) (const False)
```

Comment:
Worth a note that it's unkillable. Probably not an issue for the prototype, though.